File: packages/merkle-tree/src/test/permutation.test.ts
import { createTree, addItem, IMerkleTree, buildMerkleTree } from '../lib/merkle-tree';
import { visualizeMerkleTree } from '../lib/visualize';
import { createHashedItem } from './merkle-verify';

describe('Merkle Tree Permutation Tests', () => {
  const testFiles = ['a', 'b', 'c', 'd', 'e'];
  
  // Generate all permutations of the test files
  function generatePermutations<T>(arr: T[]): T[][] {
    if (arr.length <= 1) return [arr];
    
    const permutations: T[][] = [];
    for (let i = 0; i < arr.length; i++) {
      const current = arr[i];
      const remaining = [...arr.slice(0, i), ...arr.slice(i + 1)];
      const remainingPermutations = generatePermutations(remaining);
      
      for (const perm of remainingPermutations) {
        permutations.push([current, ...perm]);
      }
    }
    
    return permutations;
  }

  function createPermutationTree(permutation: string[]): IMerkleTree<any> {
    let tree = createTree<any>("12345678-1234-5678-9abc-123456789abc");
  
    // Add files in the permutation order
    for (const file of permutation) {
      const fileHash = createHashedItem(file);
      tree = addItem(tree, fileHash);

      // console.log(`Added file ${file} to tree:`);
      // console.log(`Tree:\n${visualizeTree(tree.root)}`);
    }
    return tree;
  }

  const permutations = generatePermutations(testFiles);
  const comparisonPermutation = permutations[0];

  // console.log(`Creating tree 1 [${comparisonPermutation.join('-')}]`);
  const comparisonTree = createPermutationTree(comparisonPermutation);

  comparisonTree.dirty = false;
  comparisonTree.merkle = buildMerkleTree(comparisonTree.sort); // Force tree rebuild.
  
  // Test that every other permutation produces the same tree as the first permutation.
  for (let index = 1; index < permutations.length; index++) {
    const permutation = permutations[index];
  
    test(`should produce same tree for permutation 1 [${comparisonPermutation.join('-')}] and ${index + 1} [${permutation.join('-')}]`, () => {
      // console.log(`Creating tree ${index + 1} [${permutation.join('-')}]`);
      const permutationTree = createPermutationTree(permutation);

      permutationTree.dirty = false;
      permutationTree.merkle = buildMerkleTree(permutationTree.sort); // Force tree rebuild.

      const root = permutationTree.merkle;
      expect(root).toBeDefined();

      const same = Buffer.compare(root!.hash, comparisonTree.merkle!.hash) === 0;
      if (!same) {
        const msg = `Permutation ${index + 1} [${permutation.join(', ')}] produces a different tree to permutation 1 [${comparisonPermutation.join(', ')}].`;
        console.log(msg);
        console.log(`Comparison:  ${comparisonTree.merkle!.hash.toString('hex')}`);
        console.log(`Permutation: ${root!.hash.toString('hex')}`);
        console.log(`Comparison:\n${visualizeMerkleTree(comparisonTree.merkle)}`);
        console.log(`Permutation:\n${visualizeMerkleTree(root)}`);

        throw new Error(msg);
      }
    });
  }
});

File: packages/merkle-tree-zig/src/test/permutation.test.zig
//
// Every insertion order produces the same tree (port of src/test/permutation.test.ts).
// (Zig: TypeScript generates one test per permutation; here one test checks all of them.)
//

const std = @import("std");
const merkle_tree_zig = @import("merkle-tree-zig");
const merkle_verify = @import("merkle-verify.zig");
const merkle_tree = merkle_tree_zig.merkle_tree;
const IMerkleTree = merkle_tree.IMerkleTree;

//
// Builds a tree by adding the files in the permutation order and builds its merkle tree.
//
fn createPermutationTree(allocator: std.mem.Allocator, permutation: []const []const u8) !IMerkleTree {
    var tree = try merkle_verify.buildTree(allocator, permutation);
    tree.dirty = false;
    tree.merkle = try merkle_tree.buildMerkleTree(allocator, tree.sort); // Force tree rebuild.
    return tree;
}

test "should produce same tree for permutation 1 [a-b-c-d-e] and N" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const permutations = try merkle_verify.generatePermutations(allocator, &.{ "a", "b", "c", "d", "e" });
    const comparisonTree = try createPermutationTree(allocator, permutations[0]);

    // Test that every other permutation produces the same tree as the first permutation.
    for (permutations[1..]) |permutation| {
        const permutationTree = try createPermutationTree(allocator, permutation);
        const root = permutationTree.merkle;
        try std.testing.expect(root != null);
        try std.testing.expectEqualSlices(u8, comparisonTree.merkle.?.hash, root.?.hash);
    }
}
